File: web/src/api/endpoints/redeem.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import type { User } from './user';

export type RedeemCodeType = 'balance' | 'monthly';

export interface RedeemCode {
    id: number;
    code: string;
    type: RedeemCodeType;
    enabled: boolean;
    used: boolean;
    balance_amount: number;
    // 兼容后端字段：monthly_limit 在 UI 中表示月卡每日额度。
    monthly_limit: number;
    monthly_days: number;
    daily_quota?: number;
    valid_days?: number;
    created_by_user_id: number;
    used_by_user_id: number;
    used_at: number;
    note?: string;
    created_at: number;
    updated_at: number;
}

export interface GenerateRedeemCodeRequest {
    type: RedeemCodeType;
    count: number;
    balance_amount?: number;
    // 兼容后端字段：生成月卡时作为每日额度提交。
    monthly_limit?: number;
    monthly_days?: number;
    daily_quota?: number;
    valid_days?: number;
    note?: string;
}

export function useRedeemCodeList() {
    return useQuery({
        queryKey: ['redeem', 'list'],
        queryFn: () => apiClient.get<RedeemCode[]>('/api/v1/redeem/list'),
        refetchInterval: 30000,
    });
}

export function useGenerateRedeemCode() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: (data: GenerateRedeemCodeRequest) => apiClient.post<RedeemCode[]>('/api/v1/redeem/generate', data),
        onSuccess: () => queryClient.invalidateQueries({ queryKey: ['redeem', 'list'] }),
    });
}

export function useUpdateRedeemCode() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: (data: { id: number; enabled: boolean; note?: string }) => apiClient.post<RedeemCode>('/api/v1/redeem/update', data),
        onSuccess: () => queryClient.invalidateQueries({ queryKey: ['redeem', 'list'] }),
    });
}

export function useDeleteRedeemCode() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: (id: number) => apiClient.delete<null>(`/api/v1/redeem/delete/${id}`),
        onSuccess: () => queryClient.invalidateQueries({ queryKey: ['redeem', 'list'] }),
    });
}

export function useRedeemCode() {
    return useMutation({
        mutationFn: (code: string) => apiClient.post<User>('/api/v1/redeem/redeem', { code }),
    });
}

File: internal/model/redeem.go
package model

type RedeemCodeType string

const (
	RedeemCodeTypeBalance RedeemCodeType = "balance"
	RedeemCodeTypeMonthly RedeemCodeType = "monthly"
)

type RedeemCode struct {
	ID              int            `json:"id" gorm:"primaryKey"`
	Code            string         `json:"code" gorm:"uniqueIndex;not null"`
	Type            RedeemCodeType `json:"type" gorm:"index;not null"`
	Enabled         bool           `json:"enabled" gorm:"default:true"`
	Used            bool           `json:"used" gorm:"default:false;index"`
	BalanceAmount   float64        `json:"balance_amount" gorm:"type:real;default:0"`
	MonthlyLimit    float64        `json:"monthly_limit" gorm:"type:real;default:0"`
	MonthlyDays     int            `json:"monthly_days" gorm:"default:0"`
	CreatedByUserID int            `json:"created_by_user_id" gorm:"index;default:0"`
	UsedByUserID    int            `json:"used_by_user_id" gorm:"index;default:0"`
	UsedAt          int64          `json:"used_at" gorm:"default:0"`
	Note            string         `json:"note,omitempty"`
	CreatedAt       int64          `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt       int64          `json:"updated_at" gorm:"autoUpdateTime"`
}

type RedeemCodeResponse struct {
	RedeemCode
	DailyQuota float64 `json:"daily_quota"`
	ValidDays  int     `json:"valid_days"`
}

type RedeemCodeGenerateRequest struct {
	Type          RedeemCodeType `json:"type"`
	Count         int            `json:"count"`
	BalanceAmount float64        `json:"balance_amount"`
	MonthlyLimit  float64        `json:"monthly_limit"`
	MonthlyDays   int            `json:"monthly_days"`
	DailyQuota    *float64       `json:"daily_quota,omitempty"`
	ValidDays     *int           `json:"valid_days,omitempty"`
	Enabled       bool           `json:"enabled"`
	Note          string         `json:"note"`
}

type RedeemCodeUpdateRequest struct {
	ID      int    `json:"id"`
	Enabled bool   `json:"enabled"`
	Note    string `json:"note"`
}

type RedeemCodeRedeemRequest struct {
	Code string `json:"code"`
}

func NewRedeemCodeResponse(code RedeemCode) RedeemCodeResponse {
	return RedeemCodeResponse{
		RedeemCode: code,
		DailyQuota: code.MonthlyLimit,
		ValidDays:  code.MonthlyDays,
	}
}

func NewRedeemCodeResponses(codes []RedeemCode) []RedeemCodeResponse {
	responses := make([]RedeemCodeResponse, 0, len(codes))
	for _, code := range codes {
		responses = append(responses, NewRedeemCodeResponse(code))
	}
	return responses
}
